### PR TITLE
Create input configuration by file

### PR DESCRIPTION
**Pull request by EhWhoAmI**
This makes it possible to modify the input using a config file, using `settings.toml`.

Issues with it:
 - Fails if you do not include all the inputs
 - Still needs a few keys, such as function keys
 - Does not function without `settings.toml`
 - `settings.toml` is not copied by default after build

### CODE REVIEW

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:216]: ButtonTuple_t, prefer using instead of typedef, and also use UserInputHandler::DeviceId (which should also be DeviceId_t) instead of int for the first one.

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:317]: A good time to use structured binding
```cpp
auto const &[device, button] = buttonMap.at(s);
```

Some comments for this area would be good too.
```
// If no "+" is found
```

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:217]: static constant is prefixed with sc_ `sc_buttonMap`.

Also a good idea to use a 'transparent comparator' by adding std::less<>, so this can be accessed using string view:
```cpp
std::map<std::string, button_tuple, std::less<> >
```

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:303]: Why use a string view when it's just copied into a string? String views work best if all you're doing is substr, which seems to be the case here.

File: bin/settings.toml
Comment (jonesmz) [inline bin/settings.toml:44]: what is "invert" ?

File: bin/settings.toml
Comment (jonesmz) [inline bin/settings.toml:104]: please always leave one newline at the end of a file.

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:248]: std::move your vector

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:201]: newline before brace

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:219]: i feel like the whole point of the EnTT system is that it gets used to store data like this?

@Capital-Asterisk ?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:315]: you're comparing std::string_view::npos for std::string:npos...

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:303]: Uhhh, don't copy your string view into a string...

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:309]: I'm a bit confused over these two: avoid auto. I thought `end` here was an iterator. Why is it named "end" when it's the location of a '+'? `start` can be moved down to the else block where it is actually used.

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:184]: This is the first time I've seen exceptions in this project. I'd need an expert opinion on this.

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [inline src/test_application/OSPMagnum.cpp:187]: Learning Rust and jonesmz's comments makes every unnecessary string allocation look evil to me. There has to be a way to not copy the string here.

File: src/test_application/OSPMagnum.cpp
Comment (Capital-Asterisk) [reply src/test_application/OSPMagnum.cpp:219]: This maps string identifiers to a device and button, as far as I know this is a proper use of a map.

File: bin/settings.toml
Comment (EhWhoAmI) [reply bin/settings.toml:44]: It's supposed to be holdable, I made a mistake while reading the code

**Comment by jonesmz** [comment]
@EhWhoAmI You'll need to resolve the conflicts before we'll be able to merge it.

**Comment by EhWhoAmI** [comment]
@jonesmz, yup, fixed those

**Review by Capital-Asterisk — APPROVED** [review]

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp]: Various old style Unix tools get annoyed and make noise about it.

Don't anger the old gods basically

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:255]: use std::map::insert_or_assign for this.

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:267]: If it's going to crash if it's not found, then just explicitly crash the program, don't rely on it crashing on it's own.

do something like `throw std::exception("some explaination");`

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [reply src/osp/UserInputHandler.cpp:255]: and you should also make two instances of the function.

one that takes the `name` parameter by const&, and the other by &&. That way we can get the speed gains of move semantics when the key is an rvalue reference.

Note that i say to do this way instead of taking the parameter by value, specifically because the key may not actually be moved from if the item already exists in the map, so we don't want to pay the cost of making a copy of the string if we don't need to.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:187]: don't catch std::out_of_range.

Catch std::exception.

Also, never catch an exception by value. If you catch an exception by value, you're causing object slicing.

Always catch by const&.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:203]: I'm sad that we have to have a manually specified list of strings like this.

but I can't think of any way to avoid it.

Can you?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:211]: I don't think that these should be getting cast to int.

Why cant you make the button tuple take the key type directly?

File: src/test_application/OSPMagnum.cpp
Comment (z-adams) [reply src/test_application/OSPMagnum.cpp:203]: Nope, I think it's best to just suck it up in this case. It's ugly but doing it programmatically means circumventing Magnum's abstraction layer which is a bad idea imo.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:211]: Is it because of the mouse keys?

Makes me wonder about two possibilities.

1) Should the keyboard and mouse have their own maps, instead of a shared map?
2) Instead of explicitly saying osp::sc_keyboard vs osp::sc_mouse, you could instead use an std::variant, and then determine if it's a keyboard or mouse mapping based on the item held by the variant.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:287]: static constexpr.

Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:287]: That being said, what's the main advantage of making this it's own variable, if it's only used in one place?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:300]: Consistant bracing please.

Personally I prefer braces to always go on the next line.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:299]: use c++17's if-initializer sytax.

File: src/test_application/OSPMagnum.cpp
Comment (z-adams) [reply src/test_application/OSPMagnum.cpp:300]: Next-line braces is how the rest of the repo is formatted so let's stick with that

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:304]: Can't this be a call to emplace_back, instead of push_back?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:302]: Why are you assuming that the item with key "str" will be found?

You should use .find(), and check that you aren't getting an iterator pointing to std::map::end before dereferrencing.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:311]: I think it's weird that you've got this while loop in the else-condition.

It seems to me, based on trying to understand the code, that you could have only the while loop, without needing to have the if/else branch.

File: src/test_application/OSPMagnum.h
Comment (jonesmz) [inline src/test_application/OSPMagnum.h:97]: I recommend putting some doxygen style documentation in the header files for functions declarations.

Comment (jonesmz) [reply src/test_application/OSPMagnum.h:97]: Additionally, it looks like this functions isn't used by this file.

Would it make sense for it to have it's own header file instead?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:203]: Noted.

**Comment by jonesmz** [comment]
I recommend one of:

- Only one commit.

Or

- Make a series of commits

1) Add the toml lib and nothing else.
2) Add the toml parsing code, but don't replace the existing config code
3) Add the toml file
4) Replace the existing config code with the parsing of the toml file

**Comment by EhWhoAmI** [comment]
I'll do one commit, because I have already squashed the commits XD

**Comment by jonesmz** [comment]
> I'll do one commit, because I have already squashed the commits XD

Sure, that's fine.

File: src/test_application/OSPMagnum.cpp
Comment (EhWhoAmI) [reply src/test_application/OSPMagnum.cpp:299]: You mean like this: `if (auto end = str.find(delim); end == std::string_view::npos) {`?

File: src/test_application/OSPMagnum.cpp
Comment (EhWhoAmI) [reply src/test_application/OSPMagnum.cpp:211]: `int` is also how `osp::ButtonVarConfig` stores the button, so I was just following what was already done.

File: src/test_application/OSPMagnum.cpp
Comment (EhWhoAmI) [reply src/test_application/OSPMagnum.cpp:203]: I did a programmatic method before this, but apparently not all devices map their button keys to ASCII, so this was the best  method. I didn't want to do this either.

File: src/test_application/OSPMagnum.cpp
Comment (z-adams) [reply src/test_application/OSPMagnum.cpp:299]: Yep

Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:299]: don't forget liberal use of `const`

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:211]: Sure.

Maybe @Capital-Asterisk or @z-adams can chime in on it.

I just don't see why casting to int is the right approach. Maybe there's a good reason.

File: src/test_application/OSPMagnum.cpp
Comment (EhWhoAmI) [reply src/test_application/OSPMagnum.cpp:287]: It's used in 2 places, and it's clearer this way.

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:258]: There is no such thing as const&&.

Comment (jonesmz) [reply src/osp/UserInputHandler.cpp:258]: Basically I was wanting you to make two functions.
```
void UserInputHandler::config_register_control(std::string const& name, bool holdable, std::vector<ButtonVarConfig> vars)
{
    m_controlConfigs.insert_or_assign(name, ButtonConfig{ std::move(vars), holdable, false, 0 });
}
void UserInputHandler::config_register_control(std::string && name, bool holdable, std::vector<ButtonVarConfig> vars)
{
    m_controlConfigs.insert_or_assign(std::move(name), ButtonConfig{ std::move(vars), holdable, false, 0 });
}
```

The reason for this is that insert_or_assign does not move from the first parameter if the item in question already exists, so we don't want to take the parameter by value.
Instead we want to take the parameter by const-& OR &&, so that the caller of the function gets to decide whether to use move or not.

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:276]: You don't need a return statement after a throw statement. The return statement will never be reached.

File: src/osp/UserInputHandler.h
Comment (jonesmz) [inline src/osp/UserInputHandler.h:299]: You do not need a version of this function that takes an initializer list.

The version that takes the std::vector is sufficent.

File: src/osp/UserInputHandler.h
Comment (jonesmz) [inline src/osp/UserInputHandler.h:302]: But remember, you need const& and &&, for the name parameter, as two seperate functions.

File: src/osp/UserInputHandler.h
Comment (jonesmz) [inline src/osp/UserInputHandler.h:349]: Don't comment things out like this. Either delete the line entirely, or fix the reason why you needed to unhide these variables.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:189]: I'm not sure I agree with just catching the exception and doing nothing. The program is probably in a bad state if an exc eption gets to this level.

Just let the exception go to the top level and crash the program

File: src/osp/UserInputHandler.cpp
Comment (jonesmz) [inline src/osp/UserInputHandler.cpp:245]: As I said in the .h file, there's no need to have a version of this function that takes an initializer list. It always gets converted to std::vector, so just keep that version(s) of the function.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:173]: If you're not going to modify, or move, the `std::string primary` then bind the temporary returned by toml::find to a const& variable, like this:
```
std::string const& primary = toml::find();.
```
The reason to get into the habit of doing this is that you don't typically know (sometimes you do, most times you dont) whether the function you are calling is returning a std::string by *value*, or is instead returning a const& to a std::string that lives somewhere else.

So if you aren't intending to modify / move the string, you don't want to make a copy of the string if the return type is const&, you just want the const&.

If the function returns by value, then the temporary value that is returned can either be stored as a real variable (what you're doing now) or it can be stored as a "temporary with lifetime extension", and then your const& points to that temporary, which is what I'm recommending you do.

Comment (jonesmz) [reply src/test_application/OSPMagnum.cpp:173]: In the generic sense, there are a few catagories of how to accept return values. But I'll give you the very simplified version.

1) You don't modify or move from the variable. It's "const" as far as the current code cares. Accpet the variable as const&.
2) You *do* modify or move the variable. Accept the variable by value, without const&.

File: src/osp/UserInputHandler.h
Comment (EhWhoAmI) [reply src/osp/UserInputHandler.h:302]: Are they not?

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:184]: This is the last place where you're referencing "controls". 
As such you'll need to make sure you use std::move() on it to tell the compiler you're done with it.
Sadly, it can't detect that itself.
See here for a bit more info: https://stackoverflow.com/questions/45763943/compiler-deduction-of-rvalue-references-for-variables-going-out-of-scope

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:179]: Here you're failing to use std::move on your osp::ButtonVarConfig objects.

Use
```
controls.insert(controls.end(), std::make_move_iterator(secondaryKeys.begin()), std::make_move_iterator(secondaryKeys.end());
```
to tell the controls variable that it should move from the secondaryKeys array, instead of copy.

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:304]: no need for a line-break here.

Also, instead of `auto const it` use `auto const& it`

File: src/test_application/OSPMagnum.cpp
Comment (jonesmz) [inline src/test_application/OSPMagnum.cpp:318]: no need for a line-break here.

Also, instead of auto const it use auto const& it

**Comment by jonesmz** [comment]
At this point, I think this PR is nearly ready to merge.

I have only a very small number of complaints, which I think you can rectify fairly easily.

As such, when you've fixed these last items, I'll approve the PR, and then we'll only need @z-adams to give a final approval.

File: src/osp/UserInputHandler.h
Comment (EhWhoAmI) [reply src/osp/UserInputHandler.h:349]: That was for debugging something, I just forgot to uncomment it.

File: src/osp/UserInputHandler.h
Comment (jonesmz) [reply src/osp/UserInputHandler.h:299]: Currently you have one function that takes const&&_ which isn't really a thing

**Review by jonesmz — APPROVED** [review]

**Pull request merged**
